### ISSUE TITLE
When getting the back buffer data with an applied render target, the result is upside down

### ISSUE THREAD

**Issue by Ellpeck** [MonoGame/MonoGame#7287]
When using `GraphicsDevice.GetBackBufferData` in DesktopGL while a `RenderTarget2D` is applied, the returned texture data is ordered incorrectly. When it is stored in a texture without re-ordering, you can see that the data is stored "upside down".

This issue does not occur when there is no applied render target.

The following two screenshots were taken in a [test app](https://gist.github.com/Ellpeck/d687f67ad623e5bbe1de63c63d670a97). The left side of each screenshot is the original drawing, and the right side is an overlay of a texture generated from getting the drawn data from the back buffer. The first screenshot was taken with a render target applied, the second one was taken without one applied.

![dotnet_CT6uXZ4dWs](https://user-images.githubusercontent.com/5741138/89740359-2dfa3080-da88-11ea-9e82-7196a3f838e4.png)
![dotnet_EHgfjw7uZq](https://user-images.githubusercontent.com/5741138/89740369-35b9d500-da88-11ea-887c-ee403d448ead.png)

#### What version of MonoGame does the bug occur on:
- MonoGame 3.8.0.1375-develop

#### What operating system are you using:
- Windows

#### What MonoGame platform are you using:

- DesktopGL

**Comment by Ellpeck** [comment]
There also seem to be some other issues with the `GetBackBufferData` method: If a render target is currently applied that is smaller than the window's size, the rectangle passed into the method yields the data from an incorrect location. It seems like this is due to `PresentationParameters.BackBufferWidth` and `BackBufferHeight` being used to calculate where the pixels are located, which don't update when a render target is applied.

**Comment by Jjagg** [comment]
The code is here: https://github.com/MonoGame/MonoGame/blob/6b33e47c0fb4ac3e0c388dd14cd96d0e1a7930a6/MonoGame.Framework/Platform/Graphics/GraphicsDevice.OpenGL.cs#L1214

The correct behavior is to always read data from the default frame buffer, even if a render target is applied. So we probably need to bind the default frame buffer (if it isn't bound) before reading from it and rebind the RT (if any) afterwards. We should document the performance impact this swapping causes.

**Comment by Ellpeck** [comment]
If the correct behavior is not reading from render targets at all, then how is that achieved? There is `Texture2D.GetData`, but because that creates some rather large arrays every time it is called (whereas GetBackBufferData only seems to create an array as large as the rectangle you pass), it's not feasible for actions that require data to be gotten more frequently. 

In my case, I am using GetBackBufferData and a "draw the entire object with a solid, unique color" effect to create pixel-perfect object selection using the mouse.

**Comment by Jjagg** [comment]
Texture2D.GetData also has overloads that take a rectangle to get data from. The GetBackBufferData API mimics the Texture2D.GetData API, but is specifically meant for the back buffer.

With regards to pixel perfect collision/selection, I'd say it's better to retrieve the hittable pixels of your objects once and store them in non-video memory. And maybe even beyond that reconsider if pixel perfect selection is appropriate because in my experience people often use it when a bounding box is easier and more user friendly.
However, if you want to do it with RT and GetData, I recommend calling GetData on the render target after unbinding it.

**Comment by Ellpeck** [comment]
As you can see here:
https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework/Platform/Graphics/Texture2D.OpenGL.cs#L240
Texture2D creates an array that is as big as the texture itself, no matter how large the rectangle you pass is. This causes it to use much more memory than GetBackBufferData, which only ever creates an array as large as the rectangle:
https://github.com/MonoGame/MonoGame/blob/6b33e47c0fb4ac3e0c388dd14cd96d0e1a7930a6/MonoGame.Framework/Platform/Graphics/GraphicsDevice.OpenGL.cs#L1223

Is this intended behavior, and can this be modified? (It might just be that `GetTexImage` doesn't have this functionality.) If this is an actual issue, I can open a separate thread for it as well.

**Comment by Jjagg** [comment]
This is an issue in the OpenGL implementation of Texture2D.GetData and we should fix it. It should use glReadPixels instead of glGetTexImage. We can use this issue to track this.

**Comment by Phyyl** [comment]
@Jjagg isn't glReadPixels only for reading the current framebuffer? There's also glGetTextureSubImage but it's only available in 4.5+

**Comment by Jjagg** [comment]
> isn't glReadPixels only for reading the current framebuffer?

Yes, that is correct. On mobile, with GLES, we create an FBO for every texture, and bind and read it when `Texture2D.GetData` is called (GLES does not support glGetTexImage). That's probably what I was thinking we should do. 

However, I don't know what the performance impact is of doing this is on full OpenGL. I think whatever we do here, we need to benchmark it  (in running games, not just microbenchmark) to make sure it does not have a significant negative impact on performance. 
A simpler optimization we can do is check if glGetTextureSubImage is available and use that if it is.
